migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API

bdhirsh opened the pull request:
Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required updating the existing macros (`TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`) to avoid variable name collisions. That currently happens when the macro is called twice with the same namespace/dispatch key, in the same translation unit. Which hasn't been an issue up until now, but is an issue for the caffe2 ops using the above macro.

I did so by appending a counter to the variable name each time one of the macros is called (using the existing `__COUNTER__` macro). That also required updating the operator analyzer regex, to correctly ignore the counter when figuring out the namespace of each op.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47322 make duplicate def() calls an error in the dispatcher
* **#47321 migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API**
* #47320 rename macro. TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY to TORCH_LIBRARY_FRAGMENT
* #47319 update legacy dispatcher registration API tests to avoid duplicate def() calls
* #47318 remove ops in the __caffe2 namespace

Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

bdhirsh committed:
migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API

[ghstack-poisoned]

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

I tried unsuccessfully to avoid creating that new macro (i.e. avoid naming collisions in the existing macro without having to pass in a unique identifier): see [this comment](for details)


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

I tried unsuccessfully to avoid creating that new macro (i.e. avoid naming collisions in the existing macro without having to pass in a unique identifier): see [this comment](for details)


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]

bdhirsh commented on tools/code_analyzer/run_analyzer.sh:
```diff
@@ -19,8 +19,8 @@ echo "Analyze: ${INPUT}"
-  -torch_library_init_pattern="^.*TORCH_LIBRARY_FRAGMENT_init_([^(]+)(\(.*)?$" \
-  -torch_library_init_pattern="^.*TORCH_LIBRARY_IMPL_init_([^(]+)_([^_]+)(\(.*)?$" \
+  -torch_library_init_pattern="^.*TORCH_LIBRARY_FRAGMENT_init_([^_]+)_\d*(\(.*)?$" \
+  -torch_library_init_pattern="^.*TORCH_LIBRARY_IMPL_init_([^_]+)_([^_]+)_\d*(\(.*)?$" \
```
updating the regex to skip the extra `_<digits>` that I added to the variable names in `TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`. Note that I didn't need to make any changes to `TORCH_LIBRARY`; it already enforces that it can't be called multiple times with the same namespace, guaranteeing that we don't have any variable name collisions.

Also, I don't think that the IMPL regex previously captured the dispatch key and the function arguments in separate match groups (it looked like they ended up getting bundled together into the second group, since `[^_]` greedily matched everything from the dispatch key name through the end of the function. I don't think that the other match groups are used anywhere (so it shouldn't affect anything), but this change should get the second match group to only match on the dispatch key name.

ezyang replied:
star and not plus, in the regex here?

bdhirsh commented on torch/library.h:
```diff
@@ -658,14 +658,23 @@ class TorchLibraryInit final {
+/// in the same translation unit.
+/// Note that the TORCH_LIBRARY variant doesn't run into this problem, because it enforces
+/// that it can only be called once for a given namespace.
+#define _TORCH_LIBRARY_FRAGMENT(ns, m, uid) \
```
I prepended an underscore to these "internal" macros: partially because I thought it made it more clear that no one should use it outside of library.h, and partially because it looks like overloading a macro with a different number of arguments requires more machinery. lmk if you think I should opt for something else.

ezyang replied:
Underscore is great and what I would have done. Thanks!

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required updating the existing macros (`TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`) to avoid variable name collisions. That currently happens when the macro is called twice with the same namespace/dispatch key, in the same translation unit. Which hasn't been an issue up until now, but is an issue for the caffe2 ops using the above macro.

I did so by appending a counter to the variable name each time one of the macros is called (using the existing `__COUNTER__` macro). That also required updating the operator analyzer regex, to correctly ignore the counter when figuring out the namespace of each op.


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required updating the existing macros (`TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`) to avoid variable name collisions. That currently happens when the macro is called twice with the same namespace/dispatch key, in the same translation unit. Which hasn't been an issue up until now, but is an issue for the caffe2 ops using the above macro.

I did so by appending a counter to the variable name each time one of the macros is called (using the existing `__COUNTER__` macro). That also required updating the operator analyzer regex, to correctly ignore the counter when figuring out the namespace of each op.


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]

ezyang reviewed (approved):
Looks great, thank you!

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required updating the existing macros (`TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`) to avoid variable name collisions. That currently happens when the macro is called twice with the same namespace/dispatch key, in the same translation unit. Which hasn't been an issue up until now, but is an issue for the caffe2 ops using the above macro.

I did so by appending a counter to the variable name each time one of the macros is called (using the existing `__COUNTER__` macro). That also required updating the operator analyzer regex, to correctly ignore the counter when figuring out the namespace of each op.


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]

bdhirsh committed:
Update on "migrate export_caffe2_op_to_c10.h macros to the new dispatcher registration API"

Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This is necessary because the macro currently calls the legacy API in such a way that it causes a duplicate `def()` (which I'm planning on making an error in a later diff).

This change also required updating the existing macros (`TORCH_LIBRARY_FRAGMENT` and `TORCH_LIBRARY_IMPL`) to avoid variable name collisions. That currently happens when the macro is called twice with the same namespace/dispatch key, in the same translation unit. Which hasn't been an issue up until now, but is an issue for the caffe2 ops using the above macro.

I did so by appending a counter to the variable name each time one of the macros is called (using the existing `__COUNTER__` macro). That also required updating the operator analyzer regex, to correctly ignore the counter when figuring out the namespace of each op.


Differential Revision: [D24714805](https://our.internmc.facebook.com/intern/diff/D24714805)

[ghstack-poisoned]
